### ISSUE TITLE
git_pillar conflicts

### ISSUE THREAD

**Issue by strocknar** [saltstack/salt#28311]
When I installed salt-master on CentOS 7 and use git_pillar for the backend, I get "Failed to checkout qa from git_pillar remote 'qa git@git.apparatus.net:beta/pepper.git': 37 conflicts prevent checkout". If I navigate to /var/cache/salt/master/git_pillar/<long string>/ and run 'git pull origin qa' then it creates a "master" branch. If I 'git checkout qa' then everything seems to work fine. 

This happens even if I remove /var/cache/salt/master/git_pillar/. Curious why we get the "37 conflicts" when there is nothing in the gitfs_pillar/<long_string> directory to begin with. 

```
Salt Version:
           Salt: 2015.8.1
 
Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.7.0
         Python: 2.7.5 (default, Jun 24 2015, 00:41:19)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 1.3.0
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: 1.4.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: 0.23.1
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist: centos 7.1.1503 Core
        machine: x86_64
        release: 3.10.0-123.el7.x86_64
         system: CentOS Linux 7.1.1503 Core
```

**Comment by peter-slovak** [comment]
The git pillar was rewritten for 2015.8, make sure you use the new configuration syntax described in https://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.git_pillar.html#git-pillar-2015-8-0-and-later . For example, this is the configuration I am using when working with 4 different environments:

```bash
ext_pillar:
  - git: 
    - master ssh://git@my.server.com/user/salt.git:
      - name: master
    - dev ssh://git@my.server.com/user/salt.git:
      - name: dev
    - production ssh://git@my.server.com/user/salt.git:
      - name: production
    - qa ssh://git@my.server.com/user/salt.git:
      - name: qa
```

**Comment by strocknar** [comment]
My gitfs configs look like this:

```
fileserver_backend:
  - git

gitfs_base: qa

gitfs_provider: pygit2

gitfs_remotes:
  - git@git.server.net:beta/pepper.git:
    - root: formulas
    - pubkey: /root/.ssh/id_rsa.pub
    - privkey: /root/.ssh/id_rsa

gitfs_root: formulas

ext_pillar:
  - git: 
    - qa git@git.server.net:beta/pepper.git:
      - env: base
      - root: pillar
      - pubkey: /root/.ssh/id_rsa.pub
      - privkey: /root/.ssh/id_rsa
```

**Comment by peter-slovak** [comment]
What about trying the `git_pillar_base: qa` master option instead of overriding the branch name with `- env: base`? I'm not exactly sure how the minion responds to that, but I think the `env` option should be used to map branches to envs other than `base`.

**Comment by strocknar** [comment]
I had to downgrade back to 6.7.

**Comment by terminalmage** [comment]
@strocknar Version 6.7 of what? :confused:

Your gitfs configuration has no effect on git_pillar.

**Comment by joshughes** [comment]
I am having the same sort of issue here and it doesn't make any sense... I did have this working at one point. 

```
2015-11-04 05:36:05,499 [salt.utils.gitfs ][ERROR   ][17773] Failed to checkout prd from git_pillar remote 'prd ssh://git@stash.foo.com/salt/salt-data.git': Reference 'refs/heads/master' not found
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 703, in checkout
    self.repo.checkout(local_ref)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/repository.py", line 281, in checkout
    from_ = self.head.shorthand
GitError: Reference 'refs/heads/master' not found
``` 

Is the first error I see... which is then followed by... 

```
2015-11-04 05:36:05,597 [salt.utils.gitfs ][ERROR   ][17734] Failed to checkout prd from git_pillar remote 'prd ssh://git@stash.foo.com/salt/salt-data.git': 11 conflicts prevent checkout
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 703, in checkout
    self.repo.checkout(local_ref)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/repository.py", line 278, in checkout
    self.checkout_tree(treeish, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/repository.py", line 232, in checkout_tree
    check_error(C.git_checkout_tree(self._repo, cptr[0], copts))
  File "/usr/local/lib/python2.7/dist-packages/pygit2/errors.py", line 56, in check_error
    raise GitError(message)
GitError: 11 conflicts prevent checkout
```

Almost like it is not actually trying to check out the branch. Since the pillars represent different environments the only conflict I can think of is if it was trying to start at master then update to the environment branch. It doesn't make any sense. 

I have cleared `/var/cache/salt/master` many times and think clearing it for a separate issue may have caused this problem to occur. Is there another cache directory to clear out to for the git pillar to do a full and complete refresh? 

I am going to try going back to the old syntax and see if I can get different results.

**Comment by joshughes** [comment]
```
2015-11-04 06:18:24,234 [salt.utils.gitfs ][ERROR   ][17999] Failed to checkout sandbox from git_pillar remote 'sandbox ssh://git@stash.foo.com/salt/salt-data.git': 9 conflicts prevent checkout
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 703, in checkout
    self.repo.checkout(local_ref)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/repository.py", line 278, in checkout
    self.checkout_tree(treeish, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/repository.py", line 232, in checkout_tree
    check_error(C.git_checkout_tree(self._repo, cptr[0], copts))
  File "/usr/local/lib/python2.7/dist-packages/pygit2/errors.py", line 56, in check_error
    raise GitError(message)
GitError: 9 conflicts prevent checkout
2015-11-04 06:18:24,235 [salt.utils.gitfs ][ERROR   ][17999] Failed to checkout dev from git_pillar remote 'dev ssh://git@stash.foo.com/salt/salt-data.git': 10 conflicts prevent checkout
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 703, in checkout
    self.repo.checkout(local_ref)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/repository.py", line 278, in checkout
    self.checkout_tree(treeish, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/repository.py", line 232, in checkout_tree
    check_error(C.git_checkout_tree(self._repo, cptr[0], copts))
  File "/usr/local/lib/python2.7/dist-packages/pygit2/errors.py", line 56, in check_error
    raise GitError(message)
GitError: 10 conflicts prevent checkout
2015-11-04 06:18:24,236 [salt.utils.gitfs ][ERROR   ][17999] Failed to checkout master from git_pillar remote 'master ssh://git@stash.foo.com/salt/salt-data.git': "The reference 'refs/remotes/origin/master' cannot be peeled - Cannot retrieve reference target"
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 697, in checkout
    oid = self.repo.lookup_reference(remote_ref).get_object().id
KeyError: "The reference 'refs/remotes/origin/master' cannot be peeled - Cannot retrieve reference target"
```

now it seems even master is broken?

```
2015-11-04 06:18:49,811 [salt.utils.gitfs ][ERROR   ][17846] Exception 'u'Object not found - failed to find pack entry (7a13a37ba7c44431b5fa5b5368b97bee45cdc091)'' caught while fetching git_pillar remote 'master ssh://git@stash.foo.com/salt/salt-data.git'
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 1702, in fetch_remotes
    if repo.fetch():
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 882, in fetch
    fetch_results = origin.fetch()
  File "/usr/local/lib/python2.7/dist-packages/pygit2/remote.py", line 237, in fetch
    check_error(err)
  File "/usr/local/lib/python2.7/dist-packages/pygit2/errors.py", line 50, in check_error
    raise KeyError(message)
KeyError: u'Object not found - failed to find pack entry (7a13a37ba7c44431b5fa5b5368b97bee45cdc091)'
```

When I go back to the old method the checkout works fine... 

```yaml
ext_pillar:
  - git: master ssh://git@stash.itriagehealth.com/salt/salt-data.git root=pillar
  - git: sandbox ssh://git@stash.itriagehealth.com/salt/salt-data.git root=pillar
```

Although with the old method things don't merge I would expect them to....

**Comment by terminalmage** [comment]
What version of pygit2 are you running? Some things appear to have changed in 0.23 which affect things like authentication, I'm wondering if this issue may also be related.

**Comment by joshughes** [comment]
We are at pygit 0.21.3
```
pygit2 (0.21.3)
```

I did have the exact same thought when this started happening... I went up to 0.23 and saw new and exciting errors... so then I went back to 0.21.3. No matter how much I clear out the cache this error occurs for any branch that is not master. 

The part that is most interesting to me is that we had this working for weeks, and then all of a sudden this  started happening. It does seem to coincide with forcefully clearing the `/var/cache/salt` directory

**Comment by DanyC97** [comment]
@terminalmage which version of pygit have you guys used to run your test before releasing it? i could use the same version as long as it keeps me out of trouble

**Comment by myrah** [comment]
Use GitPython instead of pygit2 if you are not in need of using the improved authentication features in pygit2. It fixed my issue.

**Comment by joshughes** [comment]
Unfortunately we do need ssh authentication so GitPython is not an option

**Comment by heini** [comment]
@joshughes you can use ssh authentication with gitpython. All you need is to setup a host entry in /root/.ssh/config.

**Comment by terminalmage** [comment]
After some investigation, it seems this is being caused by some API changes in pygit2 after 0.20.3. This is at least the third time I've run into an API change without a deprecation path for the old usage.

I should have a fix in the next couple days.

**Comment by terminalmage** [comment]
Fix submitted in #29072. You will need to delete the git_pillar cache directory after upgrading.

This fix should be part of 2015.8.3, or you can do one of two things to get it sooner:

1. Wait for the pull request to be merged and install Salt from the 2015.8 branch
2. [Checkout the pull request locally](https://help.github.com/articles/checking-out-pull-requests-locally/)
